### ISSUE TITLE
Frozen faked parameter with 'As'

### ISSUE THREAD

**Issue by twiernik** [AutoFixture/AutoFixture#437]
[Theory, AutoFakeIt]
    public void ExetuesNextIfDoesnttMatch(
       [Frozen(As=(typeof(OwinMiddleware)))]Fake<DummyMiddleware> next,
       TablesDataMiddleware sut)

So the problem with that test is that ~~xunit says Fake<DummyMiddleware>~~ says FreezingCustomization is not compatible with OwinMiddleware. Is there a way to fix this in core or should I just create new FakeFrozen attribute for handling this type of situations?

**Comment by moodmosaic** [comment]
Can you provide the exact compiler error message that you are getting?

**Comment by twiernik** [comment]
System.ArgumentExceptionThe type 'FakeItEasy.Fake`1[Parthenon.Demo.Server.Tests.DummyMiddleware]' cannot be frozen as 'Microsoft.Owin.OwinMiddleware' because the two types are not compatible.
    w Ploeh.AutoFixture.FreezingCustomization..ctor(Type targetType, Type registeredType)
    w Ploeh.AutoFixture.Xunit2.FrozenAttribute.GetCustomization(ParameterInfo parameter)
    w Ploeh.AutoFixture.Xunit2.AutoDataAttribute.CustomizeFixture(ParameterInfo p)
    w Ploeh.AutoFixture.Xunit2.AutoDataAttribute.GetData(MethodInfo methodUnderTest)

**Comment by moodmosaic** [comment]
This is probably not working because `DummyMiddleware` doesn't satisfy *any* of the rules shown [here](https://msdn.microsoft.com/en-us/library/system.type.isassignablefrom.aspx):

>[..] if any of the following conditions is true: 
* *c* and the current instance represent the same type.
* *c* is derived either directly or indirectly from the current instance.
* The current instance is an interface that *c* implements.
* *c* is a generic type parameter, and the current instance represents one of the constraints of *c*.
* *c* represents a value type, and the current instance represents `Nullable<c> [..]

You may also have a look at the [AutoFixture source code location where this exception can be raised](https://github.com/AutoFixture/AutoFixture/blob/abd525520fab601d46d9fd43ad02df1e32db9098/Src/AutoFixture/FreezingCustomization.cs#L53).

**Comment by twiernik** [comment]
Yep that's true. This is how I see this problem:

This is valid usage of frozen attribute ( if Bar : Foo )

```csharp
void FooBar([Frozen(As=(typeof(Foo))]Bar bar)
```
This is another valid usage of frozen attribute with one of mocking framework customization.

```csharp
void Run([Frozen]Auto<Bar> bar)
```

So I was expecting this would also be a valid ( if Bar : Foo )

```csharp
void Validate([Frozen(as=typeof(Foo))]Auto<Bar> foo)
``` 

But it's not.

If you ask me if it can be fixed I would say no, but I don't have good knowledge about autofixture code and I wanted to ask if it's possible since it would be a nice feature to have.

**Comment by ploeh** [comment]
If we'd been able to use generics and other, more type-safe constructs in attributes, it would have been easier to understand why this doesn't work, because then it wouldn't even have compiled. After all, `Fake<DummyMiddleware>` *is not* `OwinMiddleware` - you wouldn't be able to write

```C#
OwinMiddleware mw = new Fake<DummyMiddleware>();
```

That wouldn't compile.

Assuming that `DummyMiddleware` is a concrete class that derives from `OwinMiddleware`, this situation is a little tricky. If `DummyMiddleware` is abstract, it'd be much easier. This test passes when `DummyMiddleware` is abstract:

```C#
[Theory, AutoFakeIt]
public void FrozenFakeIsFrozenDependency(
    [Frozen]Fake<DummyMiddleware> next,
    TablesDataMiddleware sut)
{
    Assert.Equal(next.FakedObject, sut.Middleware);
}
```

`AutoFakeIt` is defined like this:

```C#
public class AutoFakeItAttribute : AutoDataAttribute
{
    public AutoFakeItAttribute() :
        base(new Fixture()
            .Customize(new AutoFakeItEasyCustomization())
            .Customize(new MiddlewareCustomization()))
    {}
}

public class MiddlewareCustomization : ICustomization
{
    public void Customize(IFixture fixture)
    {
        fixture.Customizations.Add(
            new TypeRelay(
                typeof(OwinMiddleware),
                typeof(DummyMiddleware)));
    }
}
```

Because of the `TypeRelay`, a request for `OwinMiddleware` will be translated to a request for `DummyMiddleware`.

When *AutoFakeItEasy* encounters an abstract class (or interface), it relays the request to a request for a `Fake` of that abstract class - in this case `Fake<DumyMiddleware>`. Because of the `[Frozen]` attribute, a single `Fake<DumyMiddleware>` value already exists, and is returned.

I'd recommend making `DummyMiddleware` abstract, if it isn't already. If it's a concrete class, then why do you need a `Fake` of it?

**Comment by twiernik** [comment]
Well it looks like I've made some false assumptions. `Fake<DummyMiddleware>` is not `OwinMiddleware` but also when I used freeze on `Fake<DummyMiddleware>` my intentions were to freeze `DummyMiddleware` not `Fake<DummyMiddleware>`. I looked at FakeItEasy customization as some Jedi mind trick 

>**me**: Hey Autofixture, freeze this `gives Fake<DummyMiddleware>`, its `DummyMiddleware` :wave:
>**Autofixture**: Indeed, it's DummyMiddleware, I'll freeze it for you.

If it ended here there would be no trouble, but.. 

>**me**: And one more thing, use it as it was `OwinMiddleware`, cause you know, it derives from it
>**Autofixture**: Your jedi mind tricks don't work on me!


Returning to the topic `DummyMiddleware` wasn't abstract because I never thought of that ;] It was a concrete class when I created `TablesDataMiddleware` instance by hand because I already freezed OwinMiddleware with null value. In most cases null is a desired value, but in that test case I wanted to overwrite that registration. `DummyMiddleware` inherits from OwinMiddlware which has a abstract method which I have interest in and FakeItEasy will allow to track its calls. I think I'll waive from using AutoData in this test case. 
Thanks You guys for clearing things up.
